stellargraph/data/explorer.py

```diff
@@ -789,3 +791,152 @@ def _check_neighbourhood_sizes(self, in_size, out_size):
+        sum_dist = np.sum(dist)
+        return dist / sum_dist
+
+    def _edge_biases(self, neighbours, time, bias_type, is_forward):
+        if bias_type is None:
+            # default to uniform random sampling
+            return None
+
+        times = [t for _, t in neighbours]
+
+        # time is None indicates we should obtain the minimum available time for t_0
+        t_0 = time if time is not None else min(times)
+
+        if bias_type == "exponential":
+            # exponential decay bias needs to be reversed if looking backwards in time
+            return self._exp_biases(times, t_0, decay=is_forward)
+        else:
+            raise ValueError("Unsupported bias type")
+
+    def _step(self, node, time, is_forward, bias_type, np_rs):
+        """
+        Perform 1 temporal step from a node. Returns None if a dead-end is reached.
+
+        """
+
+        def check_time(t):
+            # time is None indicates no temporal filtering (i.e. initial edge selection)
+            return (
+                (time is None)
+                or (is_forward and t > time)
+                or (not is_forward and t < time)
+            )
+
+        neighbours = [
+            (neighbour, t)
+            for neighbour, t in self.graph.neighbors(node, include_edge_weight=True)
+            if check_time(t)
+        ]
+
+        if neighbours:
+            biases = self._edge_biases(neighbours, time, bias_type, is_forward)
+            chosen_index = (
+                naive_weighted_choices(np_rs, biases)
+                if biases is not None
+                else np_rs.choice(len(neighbours))
+            )
+            next_node, next_time = neighbours[chosen_index]
+            return next_node, next_time
+        else:
+            return None
+
+    def _walk(self, node, length, bidirectional, initial_edge_bias, walk_bias, np_rs):
+        def step(t, is_forward, is_first=False):
+            n = walk[-1] if is_forward else walk[0]
+            bias = initial_edge_bias if is_first else walk_bias
+            result = self._step(
+                n, time=t, is_forward=is_forward, bias_type=bias, np_rs=np_rs
+            )
+
+            if result is None:
+                return None
+
+            next_node, next_time = result
+            # the first step treats times as backwards, but still appends as if forward
+            if is_forward or is_first:
+                walk.append(next_node)
+            else:
+                walk.appendleft(next_node)
+
+            return next_time
+
+        walk = deque([node])
+        forward_time = None
+        backward_time = None
+
+        while len(walk) < length:
+            if len(walk) == 1:
+                # initial edge selection
+                forward_time = step(None, is_forward=False, is_first=True)
+                if bidirectional:
+                    backward_time = forward_time
+
+            if forward_time is not None:
```

Comment:
Good catch with the `elif` vs. `if` behaviour.

Do we care about `length = 2`? If so, we'll need to add another check for `if len(walk) == length` here, otherwise the second step could add a third element.

Alternatively, and better... this `if` can just move out of the loop (rather silly of me to not do it like this to begin with...):


```python
forward_time = step(None, is_forward=False, is_first=True)
backward_time = forward_time if bidirectional else None

while len(walk) < length:
    if forward_time is not None:
        forward_time = step(forward_time, is_forward=True)

    ...
```

We may need a check for `length == 1`, but I think it's probably ok to add that check to `run`:

```python
def run(...):
    if length <= 1:
        raise ValueError("length: expected at least 2, found {length}")
```

because a walk of length `0` or `1` (or negative) is probably not very useful.